Support installing workload pack groups

dsplaisted opened the pull request:
- Supports installing workload pack group MSIs (#21741).  Creating the MSIs will be a separate update to the workload creation infrastructure in dotnet/arcade.
- Creates a `ManifestVersionUpdate` class to replace passing around tuples of 5 elements in a lot of places
- Switches to a new transaction infrastructure for workloads, to allow for more control of the ordering of messages and rollback actions, as well as making it explicit which installation methods should participate in the transaction (by taking an `ITransactionContext` parameter)

(This is rebased on top of #24545, so it also currently includes commits from that PR.)

dsplaisted committed:
Add test code to calculate workload pack bundles

dsplaisted committed:
Refactor workload transactions, and support installing pack groups

dsplaisted committed:
More workload pack group test code

gkulin committed:
trying to update print-rollback to include feature band

gkulin committed:
update rollback to include feature band

dsplaisted committed:
Switch from tuples to ManifestVersionUpdate class

gkulin committed:
adding some tests

gkulin committed:
adding test for rollback across feature band

gkulin committed:
update failing test to add unique identifier and add another test for multiple manifests

dsplaisted committed:
Pass transaction context to InstallWorkloadManifest

dsplaisted committed:
Remove NullTransactionContext

joeloff commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.PackGroup.cs:
```diff
@@ -0,0 +1,117 @@
+            var manifests = _workloadResolver.GetInstalledManifests();
+            foreach (var manifest in manifests)
+            {
+                var packGroupFile = Path.Combine(manifest.ManifestDirectory, "WorkloadPackGroups.json");
```
This is the new mapping file we're going to add to the manifest installers?

dsplaisted replied:
Correct.

joeloff commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/ManifestVersionUpdate.cs:
```diff
@@ -0,0 +1,94 @@
+        public int CompareTo(ManifestVersionUpdate? other)
+        {
+            if (other == null) return 1;
+            int ret = ManifestId.CompareTo(other.ManifestId);
```
For a version update of a manifest, shouldn't the IDs always be the same and the underlying versions be different. Should we instead throw something like `InvalidOperationException` if the IDs are different?

dsplaisted replied:
This is just implemented so that equality works correctly, so that in tests you can assert that the manifest updates were what was expected.

joeloff reviewed (commented):
I think I'm good with this. The tricky bit is going to be computing the group IDs/names, but that needs to happen in Arcade

dsplaisted committed:
Support garbage collecting workload pack groups

dsplaisted committed:
Support environment variable to opt out of workload pack groups

This can be a workaround if there are bugs in the pack groups or the installation code
